### ISSUE TITLE
Use main docker build command by default rather than buildx

### ISSUE THREAD

**Issue by LewisGaul** [gabrieldemarmiesse/python-on-whales#345]
Currently `docker.build()` is an alias to `docker.buildx.build()`, but it's not clear to me why this is since `docker build` and `docker buildx build` aren't entirely compatible and the latter still seems to be somewhat experimental?

I have hit two issues with `docker.buildx.build()`: using `FROM sha256:{image.id}` in a Dockerfile is not recognised (it is with `docker build`); and on another host my docker client does not have buildx installed (and I'm very reluctant for this to be a requirement).

Under the assumption that the main reason `docker build` is not implemented is because it's additional effort on top of supporting `docker buildx` commands I've put together a basic implementation that is working for me - the PR is #346, feel free to take it over or let me know how I can get it ready for merge. Unfortunately this is a breaking change, since the `docker.build()` API is slightly different... This does seem a more correct mirror of the CLI, so I wonder if there's a way to get it in with the appropriate changelog notices and version bump?

**Comment by gabrieldemarmiesse** [comment]
Thank you for the issue it gives us an opportunity to clarify things.

Docker is in the middle of a transition between the old builder (the one you want to use, lets call it docker build v0) and the new builder (buildx with buildkit, lets call it docker build v1).

Docker wants `docker build` to ultimately refer to buildx, and it has provided tools to users to enable the behaviour now and thus avoiding fixing breakages when the switch do happen. Those tools are for example "docker buildx install" or "DOCKER_BUILDKIT=1".

To avoid having to support two interfaces, as well as many conditional flags, I decided to adopt the future behaviour in advance. This also means that there is no way to use the builder v0 with Python-on-whales. 

While your use case is valid, I'm reluctant to break the API to allow access to the docker build v0 which will be removed in the future. The timeline is not certain but I expect it to be gone in less than two years. 

I think that your use cases are valid, but maybe fixing it in python-on-whales is not the right solution. I can count two independent problems from your description. 
1) buildkit does not support FROM sha:... 
2) buildx is not installed by default when installing docker, it must be installed separately, at least on linux unlike docker build v0. 

The best way to fix those issues would be to report them to the buildx repo. I know they want as much feature parity as possible with the old builder to incite people to switch, so they'll definitely want to fix your problems

A very relevant issue: https://github.com/moby/moby/issues/40379

**Comment by LewisGaul** [comment]
Thanks for the info, I can see your perspective. However, as a user (of docker, and hopefully this project too) I'm simply seeing an incompatibility between the default `docker` CLI behaviour and what's supported by this package. On the one hand I have an up-to-date docker client that does not support buildx at all, and on the other I have a package that only supports clients that do have it. I feel like it would be easier for people to pick up this project in the transition if there was an overlap in support!

Would you consider a compromise of providing access to the 'v0' build without breaking the API, whether via the existing `docker.buildx.build()` API or via a new `docker.legacy_build()` API?

Yes I could raise issue 1) with buildkit, but unfortunately my time is limited (I did have a look through and it seemed perhaps similar issues had already been raised, maybe I'm on a version that doesn't have the fix...), and I'm just trying to get something working with this package (which seems really good btw!).

**Comment by gabrieldemarmiesse** [comment]
I understand your problem. Indeed supporting the docker build v0 is not a lot of work, I was only afraid of breking the api.

I really like the idea of a `docker.legacy_build()`. The name also implies that it won't be around forever so planning a switch will be necessary.

I believe it will need to be well documented (a note at the top of the buildx page would be perfect), and to be consistent with the api and codebase `docker.legacy_build() ` should be an alias to `docker.image.legacy_build()`.

If you can do all this I would be happy to review and merge your pull request :)

**Comment by LewisGaul** [comment]
Ok, I'll try to find time for this. I still think it's a shame that `docker.build()` doesn't reflect `docker build` from the start, but I guess there had to be a compatibility break at some point with trying to keep this project in sync with the CLI.

**Comment by gabrieldemarmiesse** [comment]
Yes, also it's not a good UX to have a function that can have two different behaviours depending on the system. Docker does this because they need a transition on the `docker build` function. Because a lot of users have this in scripts already and they try to smooth the upgrade without changer the user's code. That's bad but they have no choice. 

We don't have a large existing user codebase with python-on-whales and docker build v0, thus, we can avoid this UX problem by having different functions, and each function does the same on every system, which will help users distribute their code

If you don't have the time to adapt your pull request, I can do it myself, I'm on holidays and maintainers have write access to PR's branches by default

**Comment by LewisGaul** [comment]
> If you don't have the time to adapt your pull request, I can do it myself, I'm on holidays and maintainers have write access to PR's branches by default

Thanks a lot, I was going to try and get it done tonight, but this frees me up to focus on using the lib instead! Hopefully once it's merged it should be easy enough to create a new PyPI release?

> I also found some ressource that might be useful: https://github.com/docker/cli/blob/master/docs/deprecated.md#legacy-builder-for-linux-images
> tldr: With docker 22.06 `docker build` v0 will be unavailable, the switch will happen way earlier than I thought.

Interesting, I'll take a look. We're currently using docker 18 a lot of the time, partly because we want to make sure we support users of our container image using this as a minimum version.

**Comment by gabrieldemarmiesse** [comment]
I made a new release, I hope this package will work for you :)
